manuals/vmdkmount.1
.Dd July 14, 2026
.Dt VMDKMOUNT 1
.Os
.Sh NAME
.Nm vmdkmount
.Nd mounts a VMware Virtual Disk (VMDK) image file
.Sh SYNOPSIS
.Nm vmdkmount
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm vmdkmount
is a utility to mount a VMware Virtual Disk (VMDK) image file
.Pp
.Nm vmdkmount
is part of the
.Nm libvmdk
package.
.Nm libvmdk
is a library to access the VMware Virtual Disk (VMDK) format
.Pp
.Ar source
a VMware Virtual Disk (VMDK) image file
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr, while vmdkmount will remain running in the foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vmdkmount image1.vmdk /mnt/vmdk
vmdkmount 20260714
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr vmdkinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvmdk/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* fill block tree
* replace grain table with block tree
* replace extent table with block tree

* vmdkmount empty array on close not on free

* add support for the change tracking file?
* make sure descriptor file data is cleaned on close - also if no data files were set!

* make libvmdk_handle_open_extent_data_files_file_io_pool work
* make libvmdk_handle_open_extent_data_files use libvmdk_handle_open_extent_data_files_file_io_pool

* update manuals
  - update vmdkinfo.1
  - add vmdkmount.1
* handle ZERO extents?
* extent file add windows support (encoding)
  - add codepage functions + tests
  - libvmdk_extent_descriptor.c:629
  - libvmdk_descriptor_file.c:852
* implement extent file current offset ?
* implement libvmdk_extent_file_read_grain_group_element_data
* get grain data by grain table
  - remove unused functions from libvmdk_extent_table after refactor
* implement libvmdk_handle_set_maximum_number_of_open_handles
* add support for zero-grain table entry support

* code clean up:
  - support and locale functions
* add glob functions?
* debug output
  - descriptor file: disk database

API
* add functions to access storage media data
* add functions to access extents
* add functions to access grains

Tools:
* vmdkinfo:
  - add flag to have test to check if all the extent files are there?
  - same for all the parent extent files?
  - add option to print descriptor file data?
  - print dirty status
  - add byte size string support

Tests:
* open close
  - add parent support
* read
  - add parent support
* seek
  - add parent support
* add integrity tests

* Test images
  - vmdk flat
  - vmdk flat split (done)
  - Monolithic sparse (done)
  - 2GB extent sparse (done)
  - cowd sparse
  - cowd sparse split
  - compressed (done)
  - (compressed) with parent (done)

20090912
* see `git log' for more recent change log
* initial version based on liblnk
* worked on implementing vmdk and cowd sparse data files



vmdktools/vmdkmount.c
/*
 * Mounts a VMware Virtual Disk (VMDK) image file.
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"
#include "vmdktools_getopt.h"
#include "vmdktools_i18n.h"
#include "vmdktools_libcerror.h"
#include "vmdktools_libclocale.h"
#include "vmdktools_libcnotify.h"
#include "vmdktools_libvmdk.h"
#include "vmdktools_output.h"
#include "vmdktools_signal.h"
#include "vmdktools_unused.h"

mount_handle_t *vmdkmount_mount_handle = NULL;
int vmdkmount_abort                    = 0;

/* Signal handler for vmdkmount
 */
void vmdkmount_signal_handler(
      vmdktools_signal_t signal VMDKTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vmdkmount_signal_handler";

	VMDKTOOLS_UNREFERENCED_PARAMETER( signal )

	vmdkmount_abort = 1;

	if( vmdkmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     vmdkmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vmdkmount to mount a VMware Virtual Disk (VMDK) image file.";

	vmdktools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr, while vmdkmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "image", "a VMware Virtual Disk (VMDK) image file" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	const system_character_t *path_prefix       = NULL;
	libvmdk_error_t *error                      = NULL;
	size_t path_prefix_size                     = 0;
	system_character_t * const *sources         = NULL;
	char *program                               = "vmdkmount";
	system_integer_t option                     = 0;
	int number_of_options                        = (int) ( sizeof( options ) / sizeof( vmdktools_option_t ) );
	int number_of_sources                       = 0;
	int result                                  = 0;
	int verbose                                 = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point             = NULL;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations vmdkmount_fuse_operations;

	system_character_t *option_extended_options = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                        = { program, NULL };
	struct fuse_args vmdkmount_fuse_arguments   = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args vmdkmount_fuse_arguments   = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *vmdkmount_fuse_channel    = NULL;
#endif
	struct fuse *vmdkmount_fuse_handle          = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS vmdkmount_dokan_operations;
	DOKAN_OPTIONS vmdkmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vmdktools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vmdktools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vmdktools_output_version_fprint(
	 stdout,
	 program );

	if( vmdktools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vmdktools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vmdktools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				vmdktools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vmdktools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source image(s).\n" );

		vmdktools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	sources           = &( argv[ optind ] );
	number_of_sources = argc - optind - 1;

	if( ( optind + 1 ) == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		vmdktools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ argc - 1 ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libvmdk_notify_set_stream(
	 stderr,
	 NULL );
	libvmdk_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &vmdkmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
#if defined( WINAPI )
	path_prefix = _SYSTEM_STRING( "\\VMDK" );
#else
	path_prefix = _SYSTEM_STRING( "/vmdk" );
#endif
	path_prefix_size = 1 + system_string_length(
	                        path_prefix );

	if( mount_handle_set_path_prefix(
	     vmdkmount_mount_handle,
	     path_prefix,
	     path_prefix_size,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set path prefix.\n" );

		goto on_error;
	}
	result = mount_handle_open(
	          vmdkmount_mount_handle,
	          sources,
	          number_of_sources,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source images(s).\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "Unsupported disk type.\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		vmdkmount_fuse_arguments.argc = 0;
		vmdkmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &vmdkmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &vmdkmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &vmdkmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &vmdkmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	vmdkmount_fuse_operations.open       = &mount_fuse_open;
	vmdkmount_fuse_operations.read       = &mount_fuse_read;
	vmdkmount_fuse_operations.release    = &mount_fuse_release;
	vmdkmount_fuse_operations.opendir    = &mount_fuse_opendir;
	vmdkmount_fuse_operations.readdir    = &mount_fuse_readdir;
	vmdkmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	vmdkmount_fuse_operations.getattr    = &mount_fuse_getattr;
	vmdkmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	vmdkmount_fuse_handle = fuse_new(
	                         &vmdkmount_fuse_arguments,
	                         &vmdkmount_fuse_operations,
	                         sizeof( struct fuse_operations ),
	                         vmdkmount_mount_handle );

	if( vmdkmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          vmdkmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	vmdkmount_fuse_channel = fuse_mount(
	                          mount_point,
	                          &vmdkmount_fuse_arguments );

	if( vmdkmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	vmdkmount_fuse_handle = fuse_new(
	                         vmdkmount_fuse_channel,
	                         &vmdkmount_fuse_arguments,
	                         &vmdkmount_fuse_operations,
	                         sizeof( struct fuse_operations ),
	                         vmdkmount_mount_handle );

	if( vmdkmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          vmdkmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 vmdkmount_fuse_handle );

	fuse_destroy(
	 vmdkmount_fuse_handle );

	fuse_opt_free_args(
	 &vmdkmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &vmdkmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &vmdkmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	vmdkmount_dokan_options.Version    = DOKAN_VERSION;
	vmdkmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	vmdkmount_dokan_options.SingleThread = TRUE;
#else
	vmdkmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		vmdkmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		vmdkmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	vmdkmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	vmdkmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	vmdkmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	vmdkmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	vmdkmount_dokan_operations.CreateDirectory      = NULL;
	vmdkmount_dokan_operations.Cleanup              = NULL;
	vmdkmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	vmdkmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	vmdkmount_dokan_operations.WriteFile            = NULL;
	vmdkmount_dokan_operations.FlushFileBuffers     = NULL;
	vmdkmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	vmdkmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	vmdkmount_dokan_operations.FindFilesWithPattern = NULL;
	vmdkmount_dokan_operations.SetFileAttributes    = NULL;
	vmdkmount_dokan_operations.SetFileTime          = NULL;
	vmdkmount_dokan_operations.DeleteFile           = NULL;
	vmdkmount_dokan_operations.DeleteDirectory      = NULL;
	vmdkmount_dokan_operations.MoveFile             = NULL;
	vmdkmount_dokan_operations.SetEndOfFile         = NULL;
	vmdkmount_dokan_operations.SetAllocationSize    = NULL;
	vmdkmount_dokan_operations.LockFile             = NULL;
	vmdkmount_dokan_operations.UnlockFile           = NULL;
	vmdkmount_dokan_operations.GetFileSecurity      = NULL;
	vmdkmount_dokan_operations.SetFileSecurity      = NULL;
	vmdkmount_dokan_operations.GetDiskFreeSpace     = NULL;
	vmdkmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	vmdkmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	vmdkmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	vmdkmount_dokan_operations.Cleanup              = NULL;
	vmdkmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	vmdkmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	vmdkmount_dokan_operations.WriteFile            = NULL;
	vmdkmount_dokan_operations.FlushFileBuffers     = NULL;
	vmdkmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	vmdkmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	vmdkmount_dokan_operations.FindFilesWithPattern = NULL;
	vmdkmount_dokan_operations.SetFileAttributes    = NULL;
	vmdkmount_dokan_operations.SetFileTime          = NULL;
	vmdkmount_dokan_operations.DeleteFile           = NULL;
	vmdkmount_dokan_operations.DeleteDirectory      = NULL;
	vmdkmount_dokan_operations.MoveFile             = NULL;
	vmdkmount_dokan_operations.SetEndOfFile         = NULL;
	vmdkmount_dokan_operations.SetAllocationSize    = NULL;
	vmdkmount_dokan_operations.LockFile             = NULL;
	vmdkmount_dokan_operations.UnlockFile           = NULL;
	vmdkmount_dokan_operations.GetFileSecurity      = NULL;
	vmdkmount_dokan_operations.SetFileSecurity      = NULL;
	vmdkmount_dokan_operations.GetDiskFreeSpace     = NULL;
	vmdkmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	vmdkmount_dokan_operations.Unmounted            = NULL;
	vmdkmount_dokan_operations.FindStreams          = NULL;
	vmdkmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &vmdkmount_dokan_options,
	          &vmdkmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount VMware Virtual Disk (VMDK) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( vmdkmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 vmdkmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &vmdkmount_fuse_arguments );
#endif
	if( vmdkmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &vmdkmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

